presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergMetadataOptimizer.java

```diff
@@ -0,0 +1,411 @@
+                DiscretePredicates predicates)
+        {
+            // Fold min/max aggregations to a constant value
+            ImmutableMap.Builder<VariableReferenceExpression, List<RowExpression>> inputColumnValuesBuilder = ImmutableMap.builder();
+            // For each input partition column, we keep one tuple domain for each constant value. When we get the resulting value, we will get the corresponding tuple domain and
+            // check if the partition stats can be trusted.
+            ImmutableMap.Builder<VariableReferenceExpression, Map<RowExpression, TupleDomain<ColumnHandle>>> inputValueToDomainBuilder = ImmutableMap.builder();
+            for (VariableReferenceExpression input : inputs) {
+                ImmutableList.Builder<RowExpression> arguments = ImmutableList.builder();
+                Map<RowExpression, TupleDomain<ColumnHandle>> valueToDomain = new HashMap<>();
+                ColumnHandle column = columns.get(input);
+                // for each input column, add a literal expression using the entry value
+                for (TupleDomain<ColumnHandle> domain : predicates.getPredicates()) {
+                    if (domain.isNone()) {
+                        continue;
+                    }
+                    Map<ColumnHandle, NullableValue> entries = TupleDomain.extractFixedValues(domain).get();
+                    NullableValue value = entries.get(column);
+                    if (value == null) {
+                        // partition key does not have a single value, so bail out to be safe
+                        return context.defaultRewrite(node);
+                    }
+                    // min/max ignores null value
+                    else if (value.getValue() != null) {
+                        Type type = input.getType();
+                        ConstantExpression constantExpression = new ConstantExpression(Optional.empty(), value.getValue(), type);
+                        arguments.add(constantExpression);
+                        valueToDomain.putIfAbsent(constantExpression, domain);
+                    }
+                }
+                inputColumnValuesBuilder.put(input, arguments.build());
+                inputValueToDomainBuilder.put(input, valueToDomain);
+            }
+            Map<VariableReferenceExpression, List<RowExpression>> inputColumnValues = inputColumnValuesBuilder.build();
+
+            Assignments.Builder assignmentsBuilder = Assignments.builder();
+            for (VariableReferenceExpression outputVariable : node.getOutputVariables()) {
+                Aggregation aggregation = node.getAggregations().get(outputVariable);
+                RowExpression inputVariable = getOnlyElement(aggregation.getArguments());
+                RowExpression result = evaluateMinMax(
+                        functionMetadataManager.getFunctionMetadata(node.getAggregations().get(outputVariable).getFunctionHandle()),
+                        inputColumnValues.get(inputVariable));
```

Comment:
Intuitively, I think it could have benefits because it further optimizes min/max to greatest/least which can do calculating in batch.

Reply:
Have you done any benchmarks to compare running the aggregations in the here in the connector vs only replacing the table scans?

Reply:
I didn't do that, but it worth a try. I will figure it out.

Reply:
I wrote a benchmark test to compare `min/max` with `least/greatest` on `ValuesNode` with different rows. The result is as follows:

```
Benchmark                                                (recordCount)  Mode  Cnt    Score    Error  Units
BenchmarkMinMaxWithLeastGreatest.baseline                        2 * 5  avgt   10   24.250 ±  0.304  ms/op
BenchmarkMinMaxWithLeastGreatest.baseline                       2 * 50  avgt   10   24.429 ±  0.383  ms/op
BenchmarkMinMaxWithLeastGreatest.baseline                      20 * 50  avgt   10   24.463 ±  0.445  ms/op
BenchmarkMinMaxWithLeastGreatest.baseline                     200 * 50  avgt   10   24.233 ±  0.382  ms/op
BenchmarkMinMaxWithLeastGreatest.testWithFurtherFlag             2 * 5  avgt   10   36.249 ±  1.835  ms/op
BenchmarkMinMaxWithLeastGreatest.testWithFurtherFlag            2 * 50  avgt   10   36.407 ±  1.333  ms/op
BenchmarkMinMaxWithLeastGreatest.testWithFurtherFlag           20 * 50  avgt   10   43.463 ±  0.607  ms/op
BenchmarkMinMaxWithLeastGreatest.testWithFurtherFlag          200 * 50  avgt   10  191.785 ±  2.027  ms/op
BenchmarkMinMaxWithLeastGreatest.testWithoutFurtherFlag          2 * 5  avgt   10   34.020 ±  0.798  ms/op
BenchmarkMinMaxWithLeastGreatest.testWithoutFurtherFlag         2 * 50  avgt   10   34.072 ±  0.894  ms/op
BenchmarkMinMaxWithLeastGreatest.testWithoutFurtherFlag        20 * 50  avgt   10   49.647 ±  1.118  ms/op
BenchmarkMinMaxWithLeastGreatest.testWithoutFurtherFlag       200 * 50  avgt   10  286.988 ± 72.066  ms/op
```

The benchmark test code could be found here: https://github.com/hantangwangd/presto/commit/9943a6ad42d5069aa4fc1045ca01c31ab75f39c0

Seems when row count is less than 1000, the two are roughly similar, with `least/greatest` have a little higher overhead. But when row count is larger, `least/greatest` would be obviously faster. The reason might be that they have the time complexity of lg(n).

Reply:
Sorry, it's just a batch execution, so roughly n/100.

Reply:
Thanks for the benchmarks! It is good to see. I think it's clear there's a benefit from these numbers and with a benefit at just 1K rows I think it's worth having.

I do have two concerns though. 

First, we don't have any limit how large the number of rows is when computing the aggregation in the optimizer. If you had a large table with 1000s of rows/partitions where you had to compute the aggregation plus many concurrent queries doing the same thing it could lead to quite a bit of CPU usage that could starve the CPU from the coordinator. I wonder if it's worth adding a config to limit the number of rows the aggregation will compute in the optimization.

2nd - Should we also limit how large the `ValuesNode` can get? The coordinator will have to ship all the data from the `ValuesNode` to a worker encoded via JSON. The data is stored in the Block form which is relatively compact, but I can see it causing GC issues on the coordinator if we have to consistently ship large `ValuesNodes` around. Additionally, those `ValuesNodes` get stored for some amount of time on the coordinator as part of the query plan's history, which could cause a large un-reclaimable memory footprint. I think it would also be good to place a limit here to prevent consuming too much memory.

Thoughts?

Reply:
Thanks for the very great questions!

I have thought about the limitation on partitions number in this optimization, and I think we should apply the limit when getting the partitions. When we get table layout for an Iceberg table we should not urgently load all the partitions there, but just return a lazily loaded data structure, it should be something like `LazyBlock`. We can loaded the data by calling `assureLoaded()` when necessary, and if it return `false` or throw an exception, we would simply skip this optimization. And when all the partitions are already loaded, we could just do the optimization. And if that's a feasible solution, then may be a subsequent separate PR would be better, what's you opinion?

I also noticed the row count of `ValuesNode` may lead in memory pressure in coordinator when checking the flaky test `AbstractTestQueries.testLargeQuery()`. And they may continuous occupying memory for a period of time as a history of the query as you described. So I agree with you that the `ValuesNode` should have a limit of row count, and maybe should use a more compacted data structure(for example use Page/Block rather than List<List<RowExpression>>), it's a problem worth to further discuss, and maybe beyond this PR's scope. If we finally decide to limit the row count of `ValuesNode`, it could be easily fitted here anyway. What do you think?

Reply:
By the way, the problems about huge `ValuesNode` make it more worthy to do further optimization by pre-calculating with `least/greatest`.

Reply:
I think we need a way to disable these optimization in this PR in the case that they cause issues on the coordinator due to memory pressure/GC issues. Either through having some flag which sets a max row count in a `ValuesNode`, or a boolean flag to disable it altogether.

Reply:
Sure, makes sense. I think a limit on row count would be better, as the queries on not so large number partitions would still get benefits when memory pressure is not so high, and we can explicitly define the value `0` as completely disable this optimization and document that.  

I think the default value should be larger than `optimizer.optimize-metadata-queries-call-threshold`'s default value 100 used in `MetadataQueryOptimizer`, for Iceberg have no need to call metastore to confirm whether each partition is valid like hive does. Maybe set to 1000, any thoughts?

Reply:
> I have thought about the limitation on partitions number in this optimization, and I think we should apply the limit when getting the partitions. When we get table layout for an Iceberg table we should not urgently load all the partitions there, but just return a lazily loaded data structure, it should be something like LazyBlock. We can loaded the data by calling assureLoaded() when necessary, and if it return false or throw an exception, we would simply skip this optimization. And when all the partitions are already loaded, we could just do the optimization. And if that's a feasible solution, then may be a subsequent separate PR would be better, what's you opinion?

I think that sounds like a reasonable solution. I am okay with waiting for a PR as long as we have a config flag added.

> I think a limit on row count for this optimization would be better than a boolean flag, as the queries on not so large number partitions would still get benefits when memory pressure is not so high, and we can explicitly define the value 0 as completely disable this optimization and document that.

I agree with this.

> I think the default value should be larger than optimizer.optimize-metadata-queries-call-threshold's default value 100 used in MetadataQueryOptimizer, for Iceberg have no need to call metastore to confirm whether each partition is valid like hive does. Maybe set to 1000, any thoughts

I think the concern about metadata calls is an adjacent issue, but one that can't be avoided from the iceberg side. Getting the list of partitions requires generating a scan plan which means loading the manifests. As long as those manifests are being cached and not requiring a additional metadata lookups, I think we don't need to worry because during actual scan planning those manifests would need to have been loaded anyways without this optimization.

The real issue is the memory-performance tradeoff. From the benchmarks we know that there is a benefit at ~1k rows, but I'm not sure how much a few ms is going to save in the grand scheme of things (at least according to your benchmark). I don't know the point at which memory usage could become an issue. I am happy to leave it at 1k for now. If you end up using this and prod and find a better value we can always tweak the default later.

Related: If we are adding configuration, we should add documentation for it as well.

Reply:
Sure, I will do this.